trackio/gpu.py

```diff
@@ -0,0 +1,322 @@
+                throttle = pynvml.nvmlDeviceGetCurrentClocksThrottleReasons(handle)
+                metrics[f"{prefix}/throttleThermal"] = int(
+                    bool(throttle & pynvml.nvmlClocksThrottleReasonSwThermalSlowdown)
+                )
+                metrics[f"{prefix}/throttlePower"] = int(
+                    bool(throttle & pynvml.nvmlClocksThrottleReasonSwPowerCap)
+                )
+                metrics[f"{prefix}/throttleHwSlowdown"] = int(
+                    bool(throttle & pynvml.nvmlClocksThrottleReasonHwSlowdown)
+                )
+                metrics[f"{prefix}/throttleApps"] = int(
+                    bool(
+                        throttle
+                        & pynvml.nvmlClocksThrottleReasonApplicationsClocksSetting
+                    )
+                )
+            except Exception:
+                pass
+
+            try:
+                ecc_corrected = pynvml.nvmlDeviceGetTotalEccErrors(
+                    handle,
+                    pynvml.NVML_MEMORY_ERROR_TYPE_CORRECTED,
+                    pynvml.NVML_VOLATILE_ECC,
+                )
+                metrics[f"{prefix}/correctedMemoryErrors"] = ecc_corrected
+            except Exception:
+                pass
+
+            try:
+                ecc_uncorrected = pynvml.nvmlDeviceGetTotalEccErrors(
+                    handle,
+                    pynvml.NVML_MEMORY_ERROR_TYPE_UNCORRECTED,
+                    pynvml.NVML_VOLATILE_ECC,
+                )
+                metrics[f"{prefix}/uncorrectedMemoryErrors"] = ecc_uncorrected
+            except Exception:
+                pass
+
+        except Exception:
+            continue
+
+    if valid_util_count > 0:
+        metrics["gpu/meanUtilization"] = total_util / valid_util_count
+    if total_mem_used_gib > 0:
+        metrics["gpu/totalMemoryGiB"] = total_mem_used_gib
+    if total_power > 0:
+        metrics["gpu/totalPowerWatts"] = total_power
+    if max_temp > 0:
+        metrics["gpu/maxTemp"] = max_temp
+
+    return metrics
+
+
+class GpuMonitor:
+    def __init__(self, run: "Run", interval: float = 10.0):
+        self._run = run
+        self._interval = interval
+        self._stop_flag = threading.Event()
+        self._thread: "threading.Thread | None" = None
+
+    def start(self):
+        if get_gpu_count() == 0:
+            warnings.warn(
+                "auto_log_gpu=True but no NVIDIA GPUs detected. GPU logging disabled."
+            )
+            return
+
+        reset_energy_baseline()
+        self._thread = threading.Thread(target=self._monitor_loop, daemon=True)
+        self._thread.start()
+
+    def stop(self):
+        self._stop_flag.set()
+        if self._thread is not None:
+            self._thread.join(timeout=2.0)
+
+    def _monitor_loop(self):
+        while not self._stop_flag.is_set():
+            try:
+                metrics = collect_gpu_metrics()
+                if metrics:
+                    self._run.log(metrics)
```

Comment:
This will increment the `step` each time a GPU metric is logged, right?

Reply:
yes that is wrong... let me see if i can attach to the current step without incrementing